# Question1: Cancer Diagnosis Using Machine Learning
## Fabio Carrasco

#### Read the dataset file “Cancer.csv” (from github using the following command), and assign it to a Pandas DataFrame:

In [13]:
import pandas as pd

In [14]:
cancer_df  = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Cancer.csv")
cancer_df.head()

Clump_Thickness  Uniformity_of_Cell_Size  Uniformity_of_Cell_Shape  \
0                5                        1                         1   
1                5                        4                         4   
2                3                        1                         1   
3                6                        8                         8   
4                4                        1                         1   

   Marginal_Adhesion  Single_Epithelial_Cell_Size  Bare_Nuclei  \
0                  1                            2            1   
1                  5                            7           10   
2                  1                            2            2   
3                  1                            3            4   
4                  3                            2            1   

   Bland_Chromatin  Normal_Nucleoli  Mitoses  Malignant_Cancer  
0                3                1        1                 0  
1                3                2        1                 0  
2                3                1        1                 0  
3                3                7        1                 0  
4                3                1        1                 0

#### Use sklearn functions to split the dataset into testing and training sets with the following parameters: test_size=0.35, random_state=3.

In [15]:
X = cancer_df.loc[:, cancer_df.columns != 'Malignant_Cancer']
y = cancer_df['Malignant_Cancer']

X.head()

Clump_Thickness  Uniformity_of_Cell_Size  Uniformity_of_Cell_Shape  \
0                5                        1                         1   
1                5                        4                         4   
2                3                        1                         1   
3                6                        8                         8   
4                4                        1                         1   

   Marginal_Adhesion  Single_Epithelial_Cell_Size  Bare_Nuclei  \
0                  1                            2            1   
1                  5                            7           10   
2                  1                            2            2   
3                  1                            3            4   
4                  3                            2            1   

   Bland_Chromatin  Normal_Nucleoli  Mitoses  
0                3                1        1  
1                3                2        1  
2                3                1        1  
3                3                7        1  
4                3                1        1

In [16]:
from sklearn.model_selection import train_test_split

In [17]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.35, 
                                                    random_state=3)

#### Use “Decision Tree Classifier” to predict Cancer based on the training/testing datasets that you built in part (b). Then, calculate and report the accuracy of your classifier.

In [18]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.utils import resample

In [19]:
my_DecisionTree = DecisionTreeClassifier(random_state=3)
my_DecisionTree.fit(X_train, y_train)
y_predict = my_DecisionTree.predict(X_test)
score_dt = accuracy_score(y_test, y_predict)
print(score_dt)

0.8301886792452831


#### The Decision Tree accuracy from part B is 83%

#### Now, we want to perform a new Ensemble Learning method called “Bagging” based on Voting on 19 decision tree classifiers.

In [20]:
predictions = []
for i in range(19):
    bootstrap_size = int(.8*len(X_train))
    X_train_i , y_train_i = resample(X_train, y_train, n_samples = bootstrap_size , random_state=i , replace = True)
    Base_DecisionTree = DecisionTreeClassifier(random_state=3)
    Base_DecisionTree.fit(X_train_i, y_train_i)
    y_predict = Base_DecisionTree.predict(X_test)
    predictions.append(y_predict)
highest_count = 0
highest_voted = predictions[0]
for i in range(19):
    one_count = list(predictions[i]).count(1)
    # print(one_count)
    if(one_count > highest_count):
        # print(one_count)
        highest_count = one_count
        highest_voted = predictions[i]
# print(elem_highest)
# print(highest_count)
score_dt = accuracy_score(y_test, highest_voted)
print(score_dt)

0.8867924528301887


#### The Final Voted accuracy is 87%

#### Use scikit-learn “Random Forest” classifier to predict Cancer based on the training/testing datasets that you built in part (b). Then, calculate and report the accuracy of your classifier. Use this command to import and define your classifier:

In [21]:
from   sklearn.ensemble    import    RandomForestClassifier

In [22]:
my_RandomForest = RandomForestClassifier(n_estimators = 19, bootstrap = True, random_state=3)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.35, 
                                                    random_state=3)
my_RandomForest.fit(X_train, y_train)
y_predict = my_RandomForest.predict(X_test)
score_dt = accuracy_score(y_test, y_predict)
print(score_dt)

0.9245283018867925


#### The Random Forest classifier generates an accuracy of 92%